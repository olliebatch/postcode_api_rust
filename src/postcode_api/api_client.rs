use surf::StatusCode;
use url::Url;

pub struct PostcodeApiClient {
    base_url: Url,
    http: surf::Client,
}

#[derive(thiserror::Error, Debug)]
pub enum PostcodeApiErrors {
    #[error("Malformed URL: {0}")]
    MalformedUrl(#[from] url::ParseError),
    #[error("Bad Response from Postcode Api client")]
    BadResponse,
    #[error("Http Client Error")]
    HttpClientError,
}

impl From<surf::Error> for PostcodeApiErrors {
    fn from(_: surf::Error) -> Self {
        Self::HttpClientError
    }
}

impl PostcodeApiClient {
    pub fn new(base_url: impl AsRef<str>) -> Result<Self, PostcodeApiErrors> {
        Ok(PostcodeApiClient {
            base_url: Url::parse(base_url.as_ref())?,
            http: surf::Client::new(),
        })
    }

    pub async fn get_post_code_info(
        self,
        postcode: String,
    ) -> Result<PostcodeClientResponse, PostcodeApiErrors> {
        let request = self
            .http
            .get(format!("{}/postcodes/{}", self.base_url, postcode));

        let mut response = request.await?;
        if response.status() != StatusCode::Ok {
            println!("{:?}", response.status());
            return Err(PostcodeApiErrors::BadResponse);
        }
        let postcode_details = response.body_json().await?;
        Ok(postcode_details)
    }
}

#[derive(Default, Debug, Clone, PartialEq, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PostcodeClientResponse {
    pub status: i64,
    pub result: PostCodeDetails,
}

#[derive(Default, Debug, Clone, PartialEq, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PostCodeDetails {
    pub postcode: String,
    pub longitude: f64,
    pub latitude: f64,
}

#[cfg(test)]
mod tests {
    use crate::postcode_api::api_client::PostcodeApiClient;
    use httpmock::MockServer;
    use serde_json::json;

    #[async_std::test]
    async fn get_postcode_details() {
        let postcode = "WC2N 5DU";
        // Arrange
        let server = MockServer::start();
        let mock = server.mock(|when, then| {
            when.method("GET").path_contains("/postcodes");
            then.status(200)
                .json_body(json!({ "status": 200, "result": {
                    "postcode": "WC2N 5DU",
                    "longitude": -0.128294,
                    "latitude": 51.507209,
                } }));
        });
        let client = PostcodeApiClient::new(&server.base_url()).unwrap();

        // Act
        let result = client
            .get_post_code_info(postcode.to_owned())
            .await
            .unwrap();

        insta::assert_json_snapshot!(result)
    }
}
