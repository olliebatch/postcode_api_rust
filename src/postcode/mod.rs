use crate::postcode_api::api_client::{PostCodeDetails, PostcodeApiClient, PostcodeApiErrors};
use serde::{Deserialize, Serialize};

// Make Location Optional as assumption that this may not come back from api with all postcodes.
#[derive(Deserialize, Serialize, Debug, Clone)]
pub struct Postcode {
    pub postcode: String,
    pub location: Option<Location>,
}

#[derive(Deserialize, Serialize, Debug, Clone)]
pub struct Location {
    latitude: f64,
    longitude: f64,
}

impl Postcode {
    pub fn new(postcode: String, location: Option<Location>) -> Self {
        Postcode { postcode, location }
    }

    pub async fn with_location(
        self,
        postcode_api: PostcodeApiClient,
    ) -> Result<Self, PostcodeApiErrors> {
        let postcode_info = postcode_api.get_post_code_info(self.postcode).await?;

        Ok(postcode_info.result.into())
    }
}

impl From<PostCodeDetails> for Postcode {
    fn from(client_response: PostCodeDetails) -> Self {
        Postcode::new(
            client_response.postcode,
            Some(Location {
                latitude: client_response.latitude,
                longitude: client_response.longitude,
            }),
        )
    }
}

#[cfg(test)]
mod tests {
    use crate::postcode::Postcode;
    use crate::postcode_api::api_client::PostcodeApiClient;
    use httpmock::MockServer;
    use serde_json::json;

    #[async_std::test]
    async fn with_location() {
        // This would be nicer to mock the trait of the profile api client.
        // For the essence of time i've left this as the same code for profile client can work here.

        let postcode = "WC2N 5DU";
        // Arrange
        let server = MockServer::start();
        let mock = server.mock(|when, then| {
            when.method("GET").path_contains("/postcodes");
            then.status(200)
                .json_body(json!({ "status": 200, "result": {
                    "postcode": "WC2N 5DU",
                    "longitude": -0.128294,
                    "latitude": 51.507209,
                } }));
        });
        let client = PostcodeApiClient::new(&server.base_url()).unwrap();

        // Act
        let postcode_struct = Postcode::new(postcode.to_owned(), None);

        let postcode_with_loc = postcode_struct.with_location(client).await.unwrap();

        insta::assert_json_snapshot!(postcode_with_loc)
    }
}
