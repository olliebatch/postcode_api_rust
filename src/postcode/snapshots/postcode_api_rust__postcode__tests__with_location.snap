---
source: src/postcode/mod.rs
expression: postcode_with_loc

---
{
  "postcode": "WC2N 5DU",
  "location": {
    "latitude": 51.507209,
    "longitude": -0.128294
  }
}
